### PR TITLE
revwalk: Allow changing hide_cb

### PR DESCRIPTION
**Pull request by TheBB**
Remove the check for existing `hide_cb` and conditionally set the `limited` flag.

Fixes https://github.com/libgit2/libgit2/issues/4887

### CODE REVIEW

**Comment by neithernut** [comment]
You may also want to test for `git_revwalk_add_hide_cb(walk, NULL, NULL)`.

**Comment by TheBB** [comment]
Sure, added a test for it.

File: tests/revwalk/hidecb.c [hunk 1]
```diff
@@ -123,7 +123,24 @@ void test_revwalk_hidecb__add_hide_cb_multiple_times(void)
 
 	cl_git_pass(git_revwalk_new(&walk, _repo));
 	cl_git_pass(git_revwalk_add_hide_cb(walk, hide_every_commit_cb, NULL));
-	cl_git_fail(git_revwalk_add_hide_cb(walk, hide_every_commit_cb, NULL));
+	cl_git_pass(git_revwalk_add_hide_cb(walk, hide_every_commit_cb, NULL));
+	cl_git_pass(git_revwalk_add_hide_cb(walk, NULL, NULL));
```
Comment (neithernut) [inline tests/revwalk/hidecb.c]: Good. But you don't verify that the revwalk actually works (e.g. by calling `git_revwalk_next()`).

Comment (TheBB) [reply tests/revwalk/hidecb.c]: Ah, gotcha. Updated again.

File: src/revwalk.c [hunk 2]
```diff
@@ -756,15 +756,11 @@ int git_revwalk_add_hide_cb(
...
-		giterr_set(GITERR_INVALID, "there is already a callback added to hide commits in revwalk");
-		return -1;
-	}
-
 	walk->hide_cb = hide_cb;
 	walk->hide_cb_payload = payload;
-	walk->limited = 1;
+
+	if (hide_cb)
+		walk->limited = 1;
```
Comment (pks-t) [inline src/revwalk.c]: Mh. We don't zero out the struct members, so you'd probably also have to unlimit the revwalk in case there is no `hide_cb`. So:

```
walk_limited = !!hide_cb;
```

File: tests/revwalk/hidecb.c [hunk 3]
```diff
@@ -123,7 +123,46 @@ void test_revwalk_hidecb__add_hide_cb_multiple_times(void)
...
+
+	git_revwalk_free(walk);
+}
+
+void test_revwalk_hidecb__unset_cb_before_walk(void)
```
Comment (pks-t) [inline tests/revwalk/hidecb.c]: This test and the previous one are actually exactly the same, except for that this new one verifies that we _did_ actually unset the hide callback. So you should probably just delete the old test

Comment (TheBB) [reply tests/revwalk/hidecb.c]: Yep, updated.

**Comment by pks-t** [comment]
Thanks a lot for this change, @TheBB!
